### PR TITLE
fix: prevent dotnet tests from reading package token

### PR DESCRIPTION
**Pull request by devantler**
> 🤖 Generated by the Agentic Engineer

## Why
Authenticated .NET restores persisted the GitHub Packages token in the user NuGet configuration, where repository-controlled tests could read it.

## What
Restore now preserves the caller NuGet configuration, adds the GitHub Packages source only for the restore step, and keeps its credential process-scoped. Tests run without restoring, and cross-platform regression coverage proves both credential isolation and caller source-mapping compatibility.

### CODE REVIEW

**Comment by devantler** [comment]
> Requested by the 🤖 Daily AI Engineer — CI is green at the current head and this draft carries no qualifying review, so requesting the lane-priority reviewer. Hygiene only: this is a sibling lane's draft and its owner promotes it.

**Comment by devantler** [comment]
@cursor review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

## User-path evaluation — c3c8a199724d42249028e5e838a178cfff59a940

- Reproduced the original credential leak: repository tests could read the persisted github credential node.
- Exercised the candidate restore with the credential scoped only to NuGet restore: all 18 existing tests passed with coverage after the credential left scope.
- Reproduced the review regression: the standalone temporary config could not restore Fixture.Dependency from a caller-mapped local source.
- Exercised the additive candidate: the mapped local package restored and its consumer test passed while the caller GitHub Packages source and package-source mapping remained effective.
- Reproduced and closed a byte-integrity edge case: NuGet add/remove rewrote the caller config BOM; the action now backs up and restores existing configs exactly.
- Reproduced the GitHub runner shell failure from an indented here-doc and replaced it with an indentation-safe source overlay; the extracted composite script passes Bash syntax and ShellCheck validation.
- Reproduced a credential-target spoof: an enabled source named github could point to a non-GitHub URL and receive the process credential. The action now fails before restore unless that source exactly matches this repository owner’s GitHub Packages endpoint; the regression fixture proves a local impostor is rejected.
- Replaced real PR credentials in the source-mapping regression with an inert sentinel while retaining coverage of the credential-bearing branch.
- Proved a pre-existing canonical caller-owned source remains byte-identical and the reusable workflow preserves the caller’s effective pre-action source state.
- Reproduced the explicit-config bypass: `RestoreConfigFile` selected a sibling NuGet config while the old overlay modified the working directory. The selected-config fixture failed before the fix, then passed restore and tests after the action overlaid the actual selected config and restored its exact SHA-256.
- Exercised both solution/default-config discovery and project/explicit-config discovery. A forced restore failure stayed nonzero and still restored the selected config bytes, including on Bash 3.2-compatible array handling.
- Reproduced the symlink data-loss path: default config discovery skipped a `NuGet.Config` symlink, cleanup deleted the link, and the target bytes changed. The action now recognizes symlinked configs; the fixture proves the link destination and target SHA-256 remain unchanged.
- Wired the source-mapping regression into both required-check aggregation surfaces; the repository’s parity guard passes locally.
- Confirmed the environment-variable name and Username/Password/ValidAuthenticationTypes format against current Microsoft NuGet documentation.
- Integrated current main, including the separately owned #672 and release 10.2.7, without modifying that excluded work.

Local yamllint and zizmor executables are unavailable. Local actionlint reports only its known unsupported code-quality permission and job.workflow_* diagnostics; GitHub CI remains the authoritative full validation.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

File: run-dotnet-tests/action.yaml [hunk 1]
```diff
@@ -21,21 +21,37 @@ runs:
...
+        <?xml version="1.0" encoding="utf-8"?>
+        <configuration>
+          <packageSources>
+            <clear />
+            <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
+            <add key="github" value="https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json" />
+          </packageSources>
+        </configuration>
+        EOF
+        dotnet restore --configfile "$config"
```
[the remark this answers is not in the document]
Comment (devantler) [reply run-dotnet-tests/action.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in 6f41b30569d550b01e674b5c0e126576e5d4671f. Authenticated restore now adds or temporarily enables the github source in the caller effective NuGet configuration, restores with process-scoped credentials, and returns the source to its prior state. A live package fixture proves an existing local source and package-source mapping still resolve.

File: run-dotnet-tests/README.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Run Dotnet Tests
 
-Test .NET solutions or projects across multiple platforms with code coverage reporting. Sets up the .NET 9 (STS) and .NET 10 (LTS) SDKs side-by-side, optionally configures GHCR for private packages, runs tests with coverage collection, and uploads the report to **GitHub Code Quality** (native PR coverage) on authenticated runs.
+Test .NET solutions or projects across multiple platforms with code coverage reporting. Sets up the .NET 9 (STS) and .NET 10 (LTS) SDKs side-by-side, optionally restores private packages from GHCR without persisting credentials, runs tests with coverage collection, and uploads the report to **GitHub Code Quality** (native PR coverage) on authenticated runs.
```
[the remark this answers is not in the document]
Comment (devantler) [reply run-dotnet-tests/README.md]: > 🤖 Generated by the Agentic Engineer

Fixed in 6f41b30569d550b01e674b5c0e126576e5d4671f. The README now names the NuGet service correctly as GitHub Packages.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

CodeRabbit updated its current-head summary with a review-limit marker and a 29-minute retry window, so the ordered review loop is advancing.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@codex review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

Codex acknowledged the request but emitted no substantive artifact during the 15-minute bounded response window, so the ordered review loop is advancing.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

**Comment by devantler** [comment]
@cursor review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

Bugbot failed before review because the account usage limit was reached, so the external-provider ladder is exhausted.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -441,6 +441,71 @@ jobs:
...
+      - name: ✅ Verify the untrusted credential target was rejected
+        env:
+          OUTCOME: ${{ steps.untrusted-source.outcome }}
+        run: |
+          if [[ "$OUTCOME" != "failure" ]]; then
+            echo "::error::Authenticated restore accepted a source named 'github' with an untrusted URL"
+            exit 1
+          fi
+        shell: bash
+
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml:583]: > 🤖 Generated by the Agentic Engineer

Fixed in de4d70e: `test-run-dotnet-tests-preserves-source-mapping` is now present in both `ci-required-checks.needs` and the aggregate `job-results`. I also executed the repository parity guard locally; it confirms the two sets are identical.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -441,6 +441,71 @@ jobs:
...
+          ../local-packages
+        shell: bash
+
+      - name: 🛡️ Reject the untrusted credential target
+        id: untrusted-source
+        continue-on-error: true
+        uses: ./run-dotnet-tests
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          working-directory: .github/fixtures/run-dotnet-tests-source-mapping/consumer
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in de4d70e. Both source-mapping invocations now use the inert literal `credential-boundary-sentinel`, so no real credential enters PR-controlled restore or test code. I intentionally kept the negative path active on pull requests rather than passing an empty token: the sentinel still exercises the credential-bearing branch and proves an impostor source is rejected, without exposing a secret.

File: .github/workflows/run-dotnet-tests.yaml [hunk 5]
```diff
@@ -91,7 +91,8 @@ jobs:
         shell: bash
         run: |
           set -euo pipefail
-          if ! dotnet nuget list source | grep -Fq 'github [Enabled]'; then
-            echo "::error::Authenticated run did not configure the GitHub Packages source"
+          if dotnet nuget list source | grep -Fq 'github [Enabled]'; then
+            echo "::error::Authenticated restore persisted the GitHub Packages source"
             exit 1
           fi
+          echo "Authenticated GitHub Packages source was removed after restore ✅"
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/run-dotnet-tests.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in de4d70e. The reusable workflow now snapshots the effective `github` source state in the caller working directory before the action and compares it afterward. The regression job separately proves that a canonical caller-owned source remains byte-identical, while the existing absent-source path still proves temporary additions are removed. The lockstep guard now requires both the trust gates and the pre/post comparison.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

Resolved the `RestoreConfigFile` finding in acd5e58. The action now evaluates the actual restore target (including projects inside a solution), overlays every distinct explicitly selected config, preserves mixed default-config projects, and restores each modified file byte-for-byte. The fixture now selects a sibling config through `RestoreConfigFile`; an MSBuild pre-restore assertion reproduced the old bypass and passes only when that selected file receives the temporary source. Local negative proof also confirms a failed restore stays nonzero while cleanup restores the selected config.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

CodeRabbit reported no additional findings, but produced only an issue comment and no review object with `commit_id` equal to the head, so it does not satisfy this repository’s green-review gate.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@codex review

File: .github/workflows/run-dotnet-tests.yaml [hunk 6]
```diff
@@ -91,7 +91,8 @@ jobs:
         shell: bash
         run: |
           set -euo pipefail
-          if ! dotnet nuget list source | grep -Fq 'github [Enabled]'; then
-            echo "::error::Authenticated run did not configure the GitHub Packages source"
+          if dotnet nuget list source | grep -Fq 'github [Enabled]'; then
+            echo "::error::Authenticated restore persisted the GitHub Packages source"
             exit 1
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/run-dotnet-tests.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in de4d70e: the reusable workflow snapshots the caller’s effective `github` source state before the action and compares the same working-directory state afterward. A canonical caller-owned source is now preserved rather than required to disappear.

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -441,6 +441,45 @@ jobs:
             exit 1
           fi
 
+  test-run-dotnet-tests-preserves-source-mapping:
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml:444]: > 🤖 Generated by the Agentic Engineer

Fixed in de4d70e: `test-run-dotnet-tests-preserves-source-mapping` is wired into both `ci-required-checks.needs` and the aggregate `job-results`. The repository parity guard passes and makes future drift fail CI.

File: run-dotnet-tests/action.yaml [hunk 8]
```diff
@@ -21,21 +21,74 @@ runs:
...
+        trap cleanup_source EXIT
+
+        source_listing="$(dotnet nuget list source --format Detailed)"
+        if grep -Eiq '^[[:space:]]*[0-9]+\.[[:space:]]+github[[:space:]]+\[Enabled\]' <<<"$source_listing"; then
+          :
+        elif grep -Eiq '^[[:space:]]*[0-9]+\.[[:space:]]+github[[:space:]]+\[Disabled\]' <<<"$source_listing"; then
+          echo "::error::The existing NuGet source named 'github' is disabled; enable or rename it before authenticated restore"
+          exit 1
+        else
+          config_path="$(find . -maxdepth 1 -type f -iname 'nuget.config' -print -quit)"
```
[the remark this answers is not in the document]
Comment (devantler) [reply run-dotnet-tests/action.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in c3c8a19. Default-config discovery now includes symlinks, so backup/add/restore follows the linked target without deleting the link. A real symlink fixture reproduced the old result (`link_present=no`, target bytes changed) and now proves both the link target and target SHA-256 remain unchanged.

File: run-dotnet-tests/action.yaml [hunk 9]
```diff
@@ -21,21 +21,74 @@ runs:
...
+              cp "$config_backup" "$config_path"
+              rm -f "$config_backup"
+              ;;
+          esac
+        }
+        trap cleanup_source EXIT
+
+        source_listing="$(dotnet nuget list source --format Detailed)"
+        if grep -Eiq '^[[:space:]]*[0-9]+\.[[:space:]]+github[[:space:]]+\[Enabled\]' <<<"$source_listing"; then
+          :
```
[the remark this answers is not in the document]
Comment (devantler) [reply run-dotnet-tests/action.yaml]: > 🤖 Generated by the Agentic Engineer

Fixed in 8eaf631 and retained through the current head. Before credential-bearing restore, an enabled source named `github` must exactly match `https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json`; otherwise the action fails. The regression fixture supplies a local impostor and proves rejection with an inert sentinel credential.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@coderabbitai review

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

CodeRabbit reported no additional findings at the full current SHA, but again emitted only an issue comment rather than a review object, so the repository’s structural green-review gate advances to Codex.

**Comment by devantler** [comment]
> 🤖 Generated by the Agentic Engineer

@codex review
